text/3541-async-drop.md

```diff
@@ -0,0 +1,222 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Native async drop introduces the `async fn drop` method as part of the `Drop` trait, enabling asynchronous cleanup during the dropping of values. While the `async_dropper` crate provides a workaround, native support ensures a more seamless integration into Rust's syntax and semantics. We can leverage async drop without relying on external crates, simplifying the codebase and reducing dependency management overhead.
+
+```rust
+use async_trait::async_trait;
+use tokio::time::{sleep, Duration};
+
+// This trait should be implement internally
+#[async_trait]
+trait AsyncDrop {
+    async fn drop(&mut self);
+}
+
+// Define an AsyncResource struct implementing AsyncDrop
+struct AsyncResource {
+    data: String,
+}
+
+// Implement AsyncDrop for AsyncResource
+#[async_trait]
+impl AsyncDrop for AsyncResource {
+    async fn drop(&mut self) {
+        println!("Async cleanup for AsyncResource: {}", self.data);
+        // Asynchronous cleanup logic, e.g., releasing resources
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    {
+        let _async_res = AsyncResource {
+            data: String::from("Example Data"),
+        };
+        // drop will be triggered here
+    }
+    // AsyncResource is dropped as it goes out of scope, and its asynchronous drop logic may be executed later by the Tokio runtime
+
+    // Sleep to ensure `drop` completed its execution
+    sleep(Duration::from_secs(2)).await;
+    println!("Main function completed");
+}
+```
+
+Moreover, to seamlessly integrate native async drop support into Rust, it is essential to outline how executors should handle asynchronous resource cleanup. This section also provides guidelines for developers using executors to ensure consistent and correct execution of async drop logic.
+
+1. **Recognition of AsyncDrop Trait**
+   Executors, such as `MyExecutor` in the following example, should be updated to recognize types implementing the `AsyncDrop` trait. This can be achieved through a mechanism such as trait bounds or associated types to identify futures that require special handling for asynchronous cleanup.
+
+   ```rust
+   trait Executor {
+       fn spawn(&self, future: impl Future);
+   }
+
+   impl<T: AsyncDrop> Executor for MyExecutor {
+       // Your custom cleanup logic goes here
+   }
+   ```
+
+   Here, `T: AsyncDrop` indicates that this executor can work with futures (`T`) that implement the AsyncDrop trait. The `AsyncDrop` trait serves as a marker, signaling that the associated type requires special handling for asynchronous cleanup.
```

Comment:
Libraries/types can't **recognize** if some other type implements a trait, they can only **require** it does, thus making the API incompatible with any type that doesn't. This is problematic because not every `Future` implements `AsyncDrop`, making this change not backward compatible.

Also, the `impl` showed here is invalid, `T` cannot be an unbound generic parameter.